python/pipeline/vreso.py

```diff
@@ -0,0 +1,175 @@
+            frame_times = patch_times[rising_edges]
+
+            # Correct NaN gaps in timestamps (mistimed or dropped packets during recording)
+            if np.any(np.isnan(frame_times)):
+                # Raise exception if first or last frame pulse was recorded in mistimed packet
+                if np.isnan(frame_times[0]) or np.isnan(frame_times[-1]):
+                    msg = ('First or last frame happened during misstamped packets. Pulses '
+                           'could have been missed: start/end of scanning is unknown.')
+                    raise PipelineException(msg)
+
+                # Fill each gap of nan values with correct number of timepoints
+                frame_period = np.nanmedian(np.diff(frame_times))  # approx
+                nan_limits = np.where(np.diff(np.isnan(frame_times)))[0]
+                nan_limits[1::2] += 1  # limits are indices of the last valid point before the nan gap and first after it
+                correct_fts = []
+                for i, (start, stop) in enumerate(zip(nan_limits[::2], nan_limits[1::2])):
+                    correct_fts.extend(frame_times[0 if i == 0 else nan_limits[2 * i - 1]: start + 1])
+                    num_missing_points = int(round((frame_times[stop] - frame_times[start]) / frame_period - 1))
+                    correct_fts.extend(np.linspace(frame_times[start], frame_times[stop], num_missing_points + 2)[1:-1])
+                correct_fts.extend(frame_times[nan_limits[-1]:])
+                frame_times = np.array(correct_fts)
+
+                # Record the NaN fix
+                num_gaps = int(len(nan_limits) / 2)
+                nan_length = sum(nan_limits[1::2] - nan_limits[::2]) * frame_period  # secs
+
+            # Check that frame times occur at the same period
+            frame_intervals = np.diff(frame_times)
+            frame_period = np.median(frame_intervals)
+            if np.any(abs(frame_intervals - frame_period) > 0.15 * frame_period):
+                raise PipelineException('Frame time period is irregular')
+
+            # Drop last frame time if scan crashed or was stopped before completion
+            valid_times = ~np.isnan(patch_times[rising_edges[0]: rising_edges[-1]])  # restricted to scan period
+            binarized_valid = binarized_signal[rising_edges[0]: rising_edges[-1]][valid_times]
+            frame_duration = np.mean(binarized_valid) * frame_period
+            falling_edges = np.where(np.diff(binarized_signal.astype(int)) < 0)[0]
+            last_frame_duration = patch_times[falling_edges[-1]] - frame_times[-1]
+            if (np.isnan(last_frame_duration) or last_frame_duration < 0 or
+                    abs(last_frame_duration - frame_duration) > 0.15 * frame_duration):
+                frame_times = frame_times[:-1]
+
+            ####### WARNING: NO CORRECTION APPLIED - TEMP CODE #######
+            voltage = f['waveform'][1, :]
```

Comment:
make sure this are float32 (or int8, int16, int32, anything other than 64 bit numbers)